rlclientlib/dedup.cc

```diff
@@ -1,83 +1,288 @@
-#include "dedup.h"
+#include "dedup_internals.h"
 #include "hash.h"
+#include "serialization/payload_serializer.h"
 #include "utility/context_helper.h"
+#include "utility/config_helper.h"
 
+#include "zstd.h"
 #include <sstream>
 
 namespace reinforcement_learning
 {
-  namespace u = utility;
+namespace u = utility;
+namespace fb = flatbuffers;
+namespace l = reinforcement_learning::logger;
 
-  dedup_dict::dict_entry::dict_entry(const char *data, size_t length) : _count(1),
-                                                                        _length(length),
-                                                                        _content(data, data + length)
+dedup_dict::dict_entry::dict_entry(const char* data, size_t length) : _count(1),
+                                                                      _length(length),
+                                                                      _content(data, data + length)
+{
+}
+
+static generic_event::object_id_t hash_content(const char*start, size_t size)
+{
+  return uniform_hash(start, size, 0);
+}
+
+generic_event::object_id_t dedup_dict::add_object(const char*start, size_t length)
```

Comment:
looks like it can be called concurrently? do we need to introduce guard here?

Reply:
I designed dedup_dict as explicitly thread unsafe.
All synchronization is done on `dedup_state`.